src/util_cs/witness_cs.rs

```diff
@@ -0,0 +1,193 @@
+//! Support for efficiently generating R1CS witness using bellperson.
+
+use ff::PrimeField;
+
+use crate::{ConstraintSystem, Index, LinearCombination, SynthesisError, Variable};
+
+pub trait SizedWitness<Scalar: PrimeField> {
+    fn num_constraints(&self) -> usize;
+    fn num_inputs(&self) -> usize;
+    fn num_aux(&self) -> usize;
+
+    fn generate_witness_into(&mut self, aux: &mut [Scalar], inputs: &mut [Scalar]) -> Scalar;
+    fn generate_witness(&mut self) -> (Vec<Scalar>, Vec<Scalar>, Scalar) {
+        let aux_count = self.num_aux();
+        let inputs_count = self.num_inputs();
+
+        let mut aux = Vec::with_capacity(aux_count);
+        let mut inputs = Vec::with_capacity(inputs_count);
+
+        aux.resize(aux_count, Scalar::ZERO);
+        inputs.resize(inputs_count, Scalar::ZERO);
+
+        let result = self.generate_witness_into(&mut aux, &mut inputs);
+
+        (aux, inputs, result)
+    }
+
+    fn generate_witness_into_cs<CS: ConstraintSystem<Scalar>>(&mut self, cs: &mut CS) -> Scalar {
+        assert!(cs.is_witness_generator());
+
+        let aux_count = self.num_aux();
+        let inputs_count = self.num_inputs();
+
+        let (aux, inputs) = cs.allocate_empty(aux_count, inputs_count);
+
+        assert_eq!(aux.len(), aux_count);
+        assert_eq!(inputs.len(), inputs_count);
+
+        self.generate_witness_into(aux, inputs)
+    }
+}
+
+/// A `ConstraintSystem` which calculates witness values for a concrete instance of an R1CS circuit.
+pub struct WitnessCS<Scalar>
+where
+    Scalar: PrimeField,
+{
+    // Assignments of variables
+    pub(crate) input_assignment: Vec<Scalar>,
+    pub(crate) aux_assignment: Vec<Scalar>,
+}
+
+use std::fmt;
+
+impl<Scalar> fmt::Debug for WitnessCS<Scalar>
+where
+    Scalar: PrimeField,
+{
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt.debug_struct("WitnessCS")
+            .field("input_assignment", &self.input_assignment)
+            .field("aux_assignment", &self.aux_assignment)
+            .finish()
+    }
+}
+
+impl<Scalar> PartialEq for WitnessCS<Scalar>
+where
+    Scalar: PrimeField,
+{
+    fn eq(&self, other: &WitnessCS<Scalar>) -> bool {
+        self.input_assignment == other.input_assignment
+            && self.aux_assignment == other.aux_assignment
+    }
+}
```

Comment:
`cargo-expand` reveals this is identical to the implementations generated by derive, is there a particular reason to not use them?

Reply:
Nope, the fields requiring it have been gradually chipped away from the original template (`ProvingAssignment` I believe) as this implementation gets moved around and modified.